api/handler/rpc/rpc.go

```diff
@@ -118,6 +118,17 @@ func (h *rpcHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	// create context
 	cx := ctx.FromRequest(r)
+	// get context from http handler wrappers
+	md, ok := r.Context().Value(metadata.MetadataKey{}).(metadata.Metadata)
+	if !ok {
+		md = make(metadata.Metadata)
+	}
+
+	// merge context with overwrite
+	cx = metadata.MergeContext(cx, md, true)
+
+	// set merged context to request
+	*r = *r.WithContext(cx)
```

Comment:
Why dereference like this?

Reply:
to modify original context inside r.
r.WithContext returns pointer to http Request. So we need to modify original pointer with new (that actually have only context differs)

Reply:
https://golang.org/src/net/http/request.go?s=13039:13098#L346

It already does this for you

Reply:
Ah crap its making a copy

Reply:
no, it returns new Request pointer (r2). Mostly i need to use *r = *r.Clone(ctx)
as i saw in docs. But non of this two functions modifies original request.